src/basic/virt.c

```diff
@@ -472,8 +465,18 @@ int detect_vm(void) {
                 return r;
         if (r == VIRTUALIZATION_VM_OTHER)
                 other = true;
-        else if (r != VIRTUALIZATION_NONE)
-                goto finish;
+        else if (r != VIRTUALIZATION_NONE) {
+                 /* If we are Dom0, then we expect to not report as a VM. However, as we might be nested
+                  * inside another hypervisor which can be detected via the CPUID check, wait to report this
+                  * until after the CPUID check. */
+                xen_dom0 = detect_vm_xen_dom0();
```

Comment:
it only makes sense to do this test if `r == VIRTUALIZATION_XEN` already applies, no? i.e. shouldn't the block be:

```c
…
else if (r == VIRTUALIZATION_XEN) {
        /* … */
        xen_dom0 = detect_vm_xen_dom0();
        if (xen_dom0 < 0)
                return xen_dom0;
        if (xen_dom0 > 0) {
                /* If we are in Xen Dom0, then we don't consider this a VM environment, because it is the domain that manages the hardware, i.e. it's the host, not the payload.
                   Don't bother with the other checks either in that case then, since we are only interested in the "innermost" VM environment, and if that's Xen Dom0, then we are not virtualized. */ 
                r = VIRTUALIZATION_NONE; 
                goto finish;
        }

        /* Otherwise: we detected Xen, and are not Dom0, hence return this as Xen */
        goto finish;
} else if (r != VIRTUALIZATION_NONE)
        goto finish;
…
```